Simplify init._calculate_fan_in_and_fan_out

JeanKossaifi opened the pull request:
This uses the shape of the tensor instead of directly indexing it. This is useful when extending PyTorch's tensor class, e.g. for lazy access. Since the `init` sub-module doesn't check for `torch_function`, it is not possibly to override its functions. Explicitly indexing the tensor will force a call to tensor() and reconstruct the full tensor/explicitly access the elements. Simply using the shape allows to avoid that.

Fixes #53540

JeanKossaifi committed:
Simplify init._calculate_fan_in_and_fan_out 

This uses the shape of the tensor instead of directly indexing it. This is useful when extending PyTorch's tensor class, e.g. for lazy access. Since the `init` sub-module doesn't check for `torch_function`, it is not possibly to override its functions. Explicitly indexing the tensor will force a call to tensor() and reconstruct the full tensor/explicitly access the elements. Simply using the shape allows to avoid that.

JeanKossaifi committed:
Provide prod for older versions

math.prod was only added in 3.8

JeanKossaifi committed:
Satisfy Lint: remove lambda + add type annotation

JeanKossaifi committed:
cosmetic: PEP8

jbschlosser commented:
Hi @JeanKossaifi ! Is there an issue related to this PR?

JeanKossaifi commented:
Hi @jbschlosser ! No, there isn't I directly made a PR, but I can also open one if you prefer.

jbschlosser commented:
> Hi @jbschlosser ! No, there isn't, I directly made a PR, but I can also open one if you prefer.

Yes please :) Please also update the description for this PR to include `Fixes #{issue number}` once that's been done.

JeanKossaifi commented:
Thanks @jbschlosser, done: issue #53540.

JeanKossaifi committed:
Manual accumulation instead of math.prod

i) math.prod is not always available
ii) reduce not supported by TorchScript
iii) such a simple case, a `prod` function is not needed

JeanKossaifi committed:
Cosmetic: flake8

jbschlosser commented on torch/nn/init.py:
```diff
@@ -274,7 +274,9 @@ def _calculate_fan_in_and_fan_out(tensor):
     receptive_field_size = 1
     if tensor.dim() > 2:
-        receptive_field_size = tensor[0][0].numel()
+        # math.prod is not always available, accumulate the product manually
```
nit: I think it's worth mentioning in this comment that `reduce` is not supported by TorchScript, as that's the true reason for the manual accumulation

JeanKossaifi replied:
Good point, will do.

jbschlosser commented on torch/nn/init.py:
```diff
@@ -274,7 +274,9 @@ def _calculate_fan_in_and_fan_out(tensor):
-        receptive_field_size = tensor[0][0].numel()
+        # math.prod is not always available, accumulate the product manually
+        for s in tensor.shape[2:]:
+            receptive_field_size = receptive_field_size * s
```
nit: `receptive_field_size *= s`

JeanKossaifi replied:
Agreed - I was worried that inplace assignment might cause issues with TorchScript.

jbschlosser reviewed (commented):
Thanks for tackling this! Couple tiny things below. Let's hope for no more unforeseen CI failures :)

JeanKossaifi committed:
Cosmetic changes
